### ISSUE TITLE
Port twisted.python.components to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5983]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5983|
|Type|enhancement|
|Created|2012-09-14 14:57:32Z|
|Branch|https://github.com/twisted/twisted/tree/components-py3-5983|

`twisted.python.components` is used e.g. by `twisted.internet.protocol`, so must be ported to Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__5983 5983
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_components_py3_5983 branches/components-py3-5983
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1347634652000000 1347634652000000
changetime__1347643260000000 1347643260000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35658]) Branching to 'components-py3-5983'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst set owner to @exarkun|
|-|-|

Ready for review. Started buildbot run:
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/components-py3-5983

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. future imports
  1. `test_addAdapter` and `test_setComponent` say they test some cache related behavior, but they don't

Looks good otherwise.  Please merge when the above are addressed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst set status to `closed`|
|-|-|

(In [35672]) Merge components-py3-5983.

Author: itamar
Review: exarkun
Fixes: [#5983](https://github.com/twisted/twisted/issues/5983)

Port twisted.python.components to Python 3, and improve test coverage.
